### PR TITLE
Remove indirection in DependencyInjection

### PR DESCRIPTION
**Pull request by pakrym**
The ServiceProviderEngine is a pretty thick abstraction and most of it didn't serve a purpose and can be inlined into the service provider removing the ICallback interface with it.

Shrink it down to only the service accessor building logic.

Remove ScopeState as it was an extra allocation per scope.

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/Expressions/ExpressionsServiceProviderEngine.cs [hunk 1]
```diff
@@ -2,23 +2,21 @@
...
         {
-            _expressionResolverBuilder = new ExpressionResolverBuilder(RuntimeResolver, this, Root);
+            _expressionResolverBuilder = new ExpressionResolverBuilder(serviceProvider);
         }
 
-        protected override Func<ServiceProviderEngineScope, object> RealizeService(ServiceCallSite callSite)
+        public override Func<ServiceProviderEngineScope, object> RealizeService(ServiceCallSite callSite)
         {
-            Func<ServiceProviderEngineScope, object> realizedService = _expressionResolverBuilder.Build(callSite);
-            RealizedServices[callSite.ServiceType] = realizedService;
```
Comment (pakrym) [inline src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/Expressions/ExpressionsServiceProviderEngine.cs:20]: These calls were redundant.

File: src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/Expressions/ExpressionResolverBuilder.cs [hunk 2]
```diff
@@ -44,39 +44,21 @@ internal sealed class ExpressionResolverBuilder : CallSiteVisitor<object, Expres
...
-            _runtimeResolver = runtimeResolver;
-            _serviceScopeFactory = serviceScopeFactory;
-            _rootScope = rootScope;
             _buildTypeDelegate = (key, cs) => BuildNoCache(cs);
         }
 
         public Func<ServiceProviderEngineScope, object> Build(ServiceCallSite callSite)
         {
-            // Optimize singleton case
-            if (callSite.Cache.Location == CallSiteResultCacheLocation.Root)
```
Comment (davidfowl) [inline src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/Expressions/ExpressionResolverBuilder.cs:74]: Why did you change this?

Comment (pakrym) [reply src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/Expressions/ExpressionResolverBuilder.cs:74]: Was able to move it way earlier into https://github.com/dotnet/runtime/pull/52140/files#diff-b9824a372eeae2c312b58b219349b8b0e4612d432c122f87b3ef45850ec2c7c2R138

**Comment by eerhardt** [comment]
@pakrym - can you resolve the conflicts on this PR?

**Comment by pakrym** [comment]
Will do.

**Comment by davidfowl** [comment]
Lets get this for preview6

File: src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/RuntimeServiceProviderEngine.cs [hunk 3]
```diff
@@ -2,24 +2,20 @@
...
-using System.Collections.Generic;
 
 namespace Microsoft.Extensions.DependencyInjection.ServiceLookup
 {
     internal sealed class RuntimeServiceProviderEngine : ServiceProviderEngine
     {
-        public RuntimeServiceProviderEngine(IEnumerable<ServiceDescriptor> serviceDescriptors) : base(serviceDescriptors)
-        {
-        }
+        public static RuntimeServiceProviderEngine Instance { get; } = new RuntimeServiceProviderEngine();
```
Comment (davidfowl) [inline src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/RuntimeServiceProviderEngine.cs:10]: I don't think we should do this, it isn't stateless right?

File: src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceProviderEngineScope.cs [hunk 4]
```diff
@@ -8,28 +8,30 @@
 
 namespace Microsoft.Extensions.DependencyInjection.ServiceLookup
 {
-    internal sealed class ServiceProviderEngineScope : IServiceScope, IServiceProvider, IAsyncDisposable
```
Comment (maryamariyan) [inline src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceProviderEngineScope.cs:11]: nit: let's rename this to `ServiceProviderScope` on a next PR

**Review by davidfowl — APPROVED** [review]

File: src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ILEmit/ILEmitResolverBuilder.cs [hunk 5]
```diff
@@ -38,39 +37,21 @@ private struct GeneratedMethod
...
 
         public Func<ServiceProviderEngineScope, object> Build(ServiceCallSite callSite)
         {
-            // Optimize singleton case
-            if (callSite.Cache.Location == CallSiteResultCacheLocation.Root)
-            {
-                object value = _runtimeResolver.Resolve(callSite, _rootScope);
-                return scope => value;
-            }
-
```
Comment (maryamariyan) [inline src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ILEmit/ILEmitResolverBuilder.cs:73]: @pakrym why is this not needed here?

Comment (maryamariyan) [reply src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ILEmit/ILEmitResolverBuilder.cs:73]: nvm, it's seems like I can use the singleton CallSiteRuntimeResolver instance here
